objects/src/notes/script.rs

```diff
@@ -1,28 +1,31 @@
-use super::{Digest, NoteError};
-use assembly::ProgramAst;
+use super::{Assembler, AssemblyContext, CodeBlock, Digest, NoteError, ProgramAst};
 
-#[derive(Debug, Eq, PartialEq)]
+#[derive(Debug, Clone)]
 pub struct NoteScript {
-    hash: Digest,
+    digest: Digest,
     code: ProgramAst,
 }
 
 impl NoteScript {
-    pub fn new<S>(script_src: S) -> Result<Self, NoteError>
-    where
-        S: AsRef<str>,
-    {
-        let code = ProgramAst::parse(script_src.as_ref()).unwrap();
-        // TODO: the code needs to be compiled with tx kernel and miden rollup library; we need
-        // to do this to get the code hash and initialize the hash filed properly
-        Ok(Self {
-            hash: Digest::default(),
-            code,
-        })
+    pub fn new(
+        code: ProgramAst,
+        assembler: &mut Assembler,
+        context: &mut AssemblyContext,
+    ) -> Result<(Self, CodeBlock), NoteError> {
+        let code_block = assembler
+            .compile_in_context(code.clone(), context)
+            .map_err(NoteError::ScriptCompilationError)?;
```

Comment:
I wonder if we should change the signature of `compile_in_context()` to accept `&ProgramAst` rather than `ProgramAst`.

Reply:
This makes sense to me.  I think we will also need to expose access methods on `ProgramAst` to return a slice over procedures and a reference to its body.  At the minute the only way to access these is to use `into_parts` which consumes the `ProgramAst`.